dataflow/src/ops/base.rs

```diff
@@ -129,6 +144,32 @@ fn key_of<'a>(key_cols: &'a [usize], r: &'a Record) -> impl Iterator<Item = &'a
         .map(move |(i, col)| key_val(i, *col, r))
 }
 
+// Replaces None values with the next available auto increment value, mutating
+// auto_increment_values if there are any changes.
+fn replace_with_auto_increment(
+    auto_increment_values: &mut HashMap<usize, DataType>,
+    row: &mut Vec<DataType>,
+    shard: Option<usize>,
+) {
+    for (index, value) in auto_increment_values.iter_mut() {
+        // When we're given a None value, replace it with the incremented version
+        // of the last auto increment value for that column (or the initial increment value):
+        *value = match (&row[*index], &value) {
+            (&DataType::None, &&mut DataType::None) => {
+                DataType::ID(shard.unwrap_or(0) as u32, INITIAL_AUTO_INCREMENT)
+            }
+            (&DataType::None, &&mut DataType::ID(s, i)) => DataType::ID(s, i + 1),
+            // Other values should override existing auto increment values, so that
+            // the auto incrementer continues from there the next time:
+            (&DataType::Int(i), &&mut DataType::ID(s, _)) => DataType::ID(s, i as i64),
+            (&DataType::BigInt(i), &&mut DataType::ID(s, _)) => DataType::ID(s, i),
+            _ => panic!("tried giving a non-numeric value to an AUTO_INCREMENT column"),
```

Comment:
This doesn't handle the case when there is no old value (i.e., `value == DataType::None`), but the new row has a value already.

Reply:
Nice catch—fixed :)